pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,287 @@
+	// if err != nil {
+	// 	return reconcile.Result{}, err
+	// }
+
+	// Get the Installation object - this is the source of truth for IP pools managed by
+	// this controller.
+	instance := &operator.Installation{}
+	if err := r.client.Get(ctx, utils.DefaultInstanceKey, instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			reqLogger.Info("Installation config not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "An error occurred when querying the Installation resource")
+		return reconcile.Result{}, err
+	}
+
+	// Mark CR found so we can report converter problems via tigerastatus
+	r.status.OnCRFound()
+
+	// SetMetaData in the TigeraStatus such as observedGenerations.
+	defer r.status.SetMetaData(&instance.ObjectMeta)
+
+	// Get all IP pools in the cluster.
+	currentPools := &crdv1.IPPoolList{}
+	err := r.client.List(ctx, currentPools)
+	if err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceReadError, "error querying IP pools", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Create a lookup map of pools owned by this controller for easy access.
+	ourPools := map[string]crdv1.IPPool{}
+	for _, p := range currentPools.Items {
+		ourPools[p.Spec.CIDR] = p
+	}
+
+	// For each pool that is desired, but doesn't exist, create it.
+	toCreate := []client.Object{}
+	for _, p := range instance.Spec.CalicoNetwork.IPPools {
+		if pool, ok := ourPools[p.CIDR]; !ok {
+			// Create a new pool.
+			toCreate = append(toCreate, p.ToCRD())
+		} else if !reflect.DeepEqual(pool, p) {
+			// Update existing pool that no longer matches.
+			toCreate = append(toCreate, p.ToCRD())
+		}
+	}
+
+	// Check existing pools owned by this controller that are no longer in
+	// the Installation resource.
+	toDelete := []client.Object{}
+	for cidr, pool := range ourPools {
+		found := false
+		for _, p := range instance.Spec.CalicoNetwork.IPPools {
+			if p.CIDR == cidr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// No match. Needs delete.
+			toDelete = append(toDelete, &pool)
+		}
+	}
+
+	// Update pools in the cluster.
+	// TODO: This is bypassing a lot of validaiton! For example:
+	// - Verify that the pool doesn't overlap with an existing pool.
+	// - Verify that the block size hasn't changed.
+	// - Verify that the pool doesn't overlap with existing IPAM blocks.
+	// We should add this validation to productize this. Ideally, we'd go through the API server (and maybe we should at steady-state).
+	// We only need to use the crd.projectcalico.org API for bootstrapping.
+	passThru := render.NewPassthrough(toCreate...)
+	handler := utils.NewComponentHandler(log, r.client, r.scheme, nil)
+	if err := handler.CreateOrUpdateOrDelete(ctx, passThru, nil); err != nil {
+		r.status.SetDegraded(operator.ResourceUpdateError, "Error creating / updating IPPools", err, log)
+		return reconcile.Result{}, err
+	}
+	delPassThru := render.NewDeletionPassthrough(toDelete...)
```

Comment:
Deletion is even more tricky - we need to disable, wait, then delete the pool, then release affinities for all the blocks within that pool. 

This is fairly complicated logic that is run within the API server today and isn't easily accessible from this controller. 

That said, there might be a way to either:
- Move that code to this controller in a maintainable way.
- Introduce a new concept of "marked for deletion", so that something like kube-controllers can actually perform the deletion and this controller would just mark the pools without needing to do the complicated parts.

Reply:
In the latest commit, I update the controller so it will only ever send deletes via the Calico API server - I think this is probably OK.

In the case that the apiserver isn't installed or functioning, it will simply disable the IP pool as a sort of stop-gap deletion until the API is available.